test/SharedAssetTest.js

```diff
@@ -0,0 +1,562 @@
+        await asset.callDecreaseShares(assetCreator, 20);
+        assert.equal(parseInt(await asset.getSharesByAddress(assetCreator)), TOTAL_SHARES - 30);
+    });
+    
+    it('transferShares transfers shares', async () => {
+        var TOTAL_SHARES = parseInt(await asset.TOTAL_SHARES());
+        assert.equal(parseInt(await asset.getSharesByAddress(assetCreator)), TOTAL_SHARES);
+        assert.equal(parseInt(await asset.getSharesByAddress(user1)), 0);
+        await asset.callTransferShares(assetCreator, user1, 10);
+        assert.equal(parseInt(await asset.getSharesByAddress(assetCreator)), TOTAL_SHARES - 10);
+        assert.equal(parseInt(await asset.getSharesByAddress(user1)), 10);
+        await asset.callTransferShares(assetCreator, user1, 20);
+        assert.equal(parseInt(await asset.getSharesByAddress(assetCreator)), TOTAL_SHARES - 30);
+        assert.equal(parseInt(await asset.getSharesByAddress(user1)), 30);
+    });
+    
+    it('addOwner/removeOwner adds/removes an owner with the given number of shares', async () => {
+        
+        // Add user1 as owner.
+        assert.equal(parseInt(await asset.getSharesByAddress(user1)), 0);
+        assert.equal(parseInt(await asset.getOwnersCount()), 1);
+        await asset.callAddOwner(user1, 10);
+        assert.equal(parseInt(await asset.getSharesByAddress(user1)), 10);
+        assert.equal(parseInt(await asset.getOwnersCount()), 2);
+        
+        // Remove user1 as owner. This does not remove the owner's shares (internally only called when the owner has 0 shares).
+        await asset.callRemoveOwner(user1);
+        assert.equal(parseInt(await asset.getOwnersCount()), 1);
+    });
+    
+    /*
+     * External payment and payout tests.
+     */
+    it('External payments are distributed over the funds and treasury', async () => {
+        
+        // Calculate expected values.
+        var payment = 1000;
+        var treasuryRatioNum = parseInt(await asset.getTreasuryRatio());
+        var treasuryRatioDenom = parseInt(await asset.TREASURY_RATIO_DENOMINATOR());
+        var treasury = Math.floor(payment * treasuryRatioNum / treasuryRatioDenom);
+        
+        // Assert actual values.
+        await asset.payment("test", {from: user1, value: 1000});
+        assert.equal(parseInt(await asset.getTreasuryBalance()), treasury);
+        assert.equal(parseInt(await web3.eth.getBalance(asset.address)), payment);
+    });
+    
+//    it('Payout to owners', async () => {
+//        
+//        // Deposit money into the treasury.
+//        await asset.treasuryDeposit('InfoMessage', {from: user1, value: 1e12});
+//        
+//        // Payout to the shareholders.
+//        var currentTimeSec = Date.now() / 1000;
+//        await asset.payOwners({timestamp: currentTimeSec + parseInt(await asset.payoutPeriod)});
+//        
+//        // Assert that the owners have received the payout.
+//        // TODO - How to get Ether from address?
```

Comment:
use [web3.eth.getBalance(address)](https://ethereum.stackexchange.com/questions/58482/getting-hold-of-web3-from-inside-a-truffle-js-test)
please validate balances after payments and share buying/selling

Reply:
Re-implemented payout test and asserted that the seller is being paid when a buyer buys their offer. If a sell offer autocompletes, it's tough to assert the balance of the seller, as the gas price creates a quite large offset here.